### PR TITLE
async memfs and async fuse

### PR DESCRIPTION
**Pull request by pwang7**
I tested this async memfs on Linux with ARM64.
The basic filesystem functionalities work, but not quiet stable yet.
I also made some fixes to async fuse.

### CODE REVIEW

File: async_fuse/src/fs/dir.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    #[test]
+    fn test_dir() -> nix::Result<()> {
+        smol::run(async {
+            let oflags = OFlag::O_RDONLY | OFlag::O_DIRECTORY;
+            let fd = blocking!(fcntl::open(".", oflags, Mode::empty()))?;
+            let dir = Dir::from_fd(fd)?;
+            let mut dir = smol::iter(dir);
+
+            while let Some(entry) = dir.next().await {
```
Comment (rogercloud) [inline async_fuse/src/fs/dir.rs:137]: next is a regular function, should we await it?

File: async_fuse/src/fs/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1420 @@
...
+        cache.insert(FUSE_ROOT_ID, root_inode);
+        let trash = BTreeSet::new(); // for deferred deletion
+
+        Ok(FileSystem { cache, trash })
+    }
+
+    /// Initialize filesystem.
+    /// Called before any other filesystem method.
+    pub fn init(&mut self, _req: &Request<'_>) -> anyhow::Result<()> {
+        Ok(())
```
Comment (rogercloud) [inline async_fuse/src/fs/mod.rs:283]: Is this function empty on purpose?

File: async_fuse/src/fs/mod.rs [hunk 3]
```diff
@@ -0,0 +1,1420 @@
...
+    }
+
+    /// Clean up filesystem.
+    /// Called on filesystem exit.
+    pub fn destroy(&mut self, _req: &Request<'_>) {}
```
Comment (rogercloud) [inline async_fuse/src/fs/mod.rs:288]: the same comment as above.

File: async_fuse/src/fs/util.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    pub rdev: u32,
+    /// Flags (macOS only, see chflags(2))
+    pub flags: u32,
+}
+
+pub fn parse_oflag(flags: u32) -> OFlag {
+    debug_assert!(
+        flags < std::i32::MAX as u32,
+        format!(
+            "helper_parse_oflag() found flags={} overflow, larger than u16::MAX",
```
Comment (rogercloud) [inline async_fuse/src/fs/util.rs:49]: the log should be "larger than i32::MAX"

File: async_fuse/src/fs/dir.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+#[cfg(not(target_os = "linux"))]
+use libc::{dirent, readdir_r};
+
+#[derive(Debug)]
+pub(crate) struct Dir(ptr::NonNull<libc::DIR>);
+
+impl Dir {
+    /// Converts from a file descriptor, closing it on success or failure.
+    pub fn from_fd(fd: RawFd) -> nix::Result<Self> {
+        let d = unsafe { libc::fdopendir(fd) };
```
[the remark this answers is not in the document]
Comment (pwang7) [reply async_fuse/src/fs/dir.rs]: I did use nix::Dir before, but nix::Dir::iter() uses Dir as borrowed reference, which is not friendly for async.
That's why I implement Dir by myself.

File: async_fuse/src/fs/mod.rs (hunk 2 above)
Comment (pwang7) [reply async_fuse/src/fs/mod.rs]: This function can be removed.

**Pull request merged**